notebooks/visit_data.py

```diff
@@ -0,0 +1,255 @@
+fields_of_interest = ['fullVisitorId', 'visitId', 'hour', 'minute', 'pagePath', 'pagePath_parent']
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# find the child page
+fields_of_interest = ['fullVisitorId', 'visitId', 'pagePath_parent', 'pagePath_child']
+visit_data_reduced['pagePath_child'] = [re.split(r'/| |\?|#', p)[-1] for p in visit_data_reduced['pagePath_clean']]
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# concat the parent and child
+pagePath_abbreviated = []
+for parent, child in zip(visit_data_reduced['pagePath_parent'], visit_data_reduced['pagePath_child']):
+    if parent != child:
+        pagePath_abbreviated.append(parent + '/' + child)
+    else:
+        pagePath_abbreviated.append(parent)
+
+visit_data_reduced['pagePath_abbreviated'] = pagePath_abbreviated
+fields_of_interest = ['fullVisitorId', 'visitId', 'pagePath_abbreviated']
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# search mask, removes about 4%
+search_mask = visit_data_reduced['pagePath_parent'].str.contains('search')
+visit_data_reduced = visit_data_reduced[~search_mask]
+# empty mask, removes 3.4%
+empty_mask = visit_data_reduced['pagePath_parent'] == ''
+visit_data_reduced = visit_data_reduced[~empty_mask]
+# sign in mask, removes 3.8%
+sign_in_mask = visit_data_reduced['pagePath'].str.contains('sign-in')
+visit_data_reduced = visit_data_reduced[~sign_in_mask]
+# login mask, removes 0.2%
+login_mask = visit_data_reduced['pagePath'].str.contains('login')
+visit_data_reduced = visit_data_reduced[~login_mask]
+# log-in mask, removes 3.2%
+log_in_mask = visit_data_reduced['pagePath'].str.contains('log-in')
+visit_data_reduced = visit_data_reduced[~log_in_mask]
+# done mask, removes 1.0%
+done_mask = visit_data_reduced['pagePath_parent'] == 'done'
+visit_data_reduced = visit_data_reduced[~done_mask]
+
+fields_of_interest = ['fullVisitorId', 'visitId', 'pagePath', 'pagePath_abbreviated']
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# lag1 page mask, removes 12%
+visit_data_reduced['pagePath_abbreviated_lag1'] = visit_data_reduced['pagePath_abbreviated'].shift()
+page_mask = visit_data_reduced['pagePath_abbreviated'] != visit_data_reduced['pagePath_abbreviated_lag1']
+visit_data_reduced = visit_data_reduced[page_mask]
+visit_data_reduced.shape
+# create a lag of visitor id to ensure that the lag1 page path relates to the same visitor
+# this cuts off the first stage of all journeys (as they have no valid lag 1 page path)
+# reduces the dataset down by a further 28%
+visit_data_reduced['fullVisitorId_lag1'] = visit_data_reduced['fullVisitorId'].shift()
+id_mask = visit_data_reduced['fullVisitorId'] == visit_data_reduced['fullVisitorId_lag1']
+visit_data_reduced = visit_data_reduced[id_mask]
```

Comment:
Are we inadvertently removing the first page a person, `fullVisitorId`, access on their journey? This is because their `fullVisitorId_lag1` will be different to their `fullVisitorId`.

Reply:
The first page are brought through to the second row of a fullVisitorId set of records, as the lagged value. The first page row itself cannot pull through a lagged version, by definition, will have a Nan there, as the user was not on Gov.UK prior to this (at least there is no data for them in this dataset) so that record is removed.

Reply:
True, that's a good point - okay with that then!